### ISSUE TITLE
Solving magneto quasi-static problem in human body models

### ISSUE THREAD

**Issue by Maatinski** [mfem/mfem#1432]
Hello MFEM community,
I've been reading through the MFEM documentation, the examples and this forum, and first of all thanks to the developers and the community for putting together this library and the knowledge base! Very impressive!!
I would like to use MFEM to simulate electric fields induced in human body models by a coil with low-frequency (LF) oscillating current (on the order of kHz). There is a few software packages available (CST, Sim4Life), but these can become pretty tedious. Our full body models are quite detailed (1mm3 hexahedral elements) resulting in something like 260M hexahedral mesh elements when using the full model. I should mention that I do have C++ experience (mostly worked with Matlab and Python though) and FEM experience, but never worked with such an advanced library, but I am eager to learn.
I am still looking through the examples and getting familiar with the MFEM library but I felt it would be good to mention what I have in mind, so that you can potentially give a few guidelines/hints on how to best achieve this.

**Magneto quasi-static problem (MQS):**
In the MQS approximation, the magnetic and electric fields are decoupled from one another (only valid for LF), i.e., the coil current drives the magnetic potential <img src="https://render.githubusercontent.com/render/math?math=\textbf{A}_0(\vec r) "> via
<img src="https://render.githubusercontent.com/render/math?math=\textbf{A}_0(\vec r) = \frac{\mu_0}{4\pi}\int_\Omega\frac{\textbf{J}_0(\vec r')}{\|\vec r - \vec r'\|} d^{3}\vec r'">

which then in turn drives the electric scalar potential <img src="https://render.githubusercontent.com/render/math?math=\varphi   "> via

<img src="https://render.githubusercontent.com/render/math?math=\nabla\cdot\sigma\nabla\varphi   = -j\omega\nabla\cdot(\sigma\textbf{A}_0)"> 
which then gives the E-field 
<img src="https://render.githubusercontent.com/render/math?math=\E = -\nabla\varphi   \label{gradphi}">

This expression is very similar to the Volta example:
<img src="https://render.githubusercontent.com/render/math?math=-\nabla\cdot\epsilon\nabla\varphi = \rho - \nabla\cdot P">
when replacing the permittivity by the conductivity, neglecting the charge density, and replacing the polarization density with the conductivity weighted magnetic potential).

**Approach:**
My idea was to apply the following changes to the Volta code to use it for simulation of magnetically induced electric fields:

1. Import a 3D coil curve that defines <img src="https://render.githubusercontent.com/render/math?math=\textbf{J}_0(\vec r)"> and use it to precompute the magnetic potential <img src="https://render.githubusercontent.com/render/math?math=\textbf{A}_0(\vec r)">
2. Ensure that the mesh elements are labelled by the correct conductivity values (we already have very detailed and correctly labeled hexahedral and tetrahedral body models)
3. Solve for the electric potential and derive the electric field.

**Questions:**
Is this workflow feasible, or am I missing something? Is there an alternative approach to solve this problem?
Ideally, I would also like to eventually compare this to the full-wave solution (e.g., using the hertz-solver) and also compare this to commercial solutions (and produce some nice visualizations for the MFEM gallery)

Thanks a lot in advance for your help!
Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
Thanks for your interest in MFEM.  It sounds like you have a very interesting project and we'e welcome your feedback (and gallery images of course!)

I'm not very familiar with this approximation although it looks very close to the standard eddy current formulation.  The only difference is that your electric field will be curl free.  This makes me a little suspicious but I'll trust that this is valid in your situation.

Your steps 2 and 3 look good. The idea of following the `Volta` miniapp to accomplish these steps should work just fine.  Step 1, however, is often non-trivial.  If you can compute this integral with pencil and paper you should consider doing that.  If not, I'd advise using the differential form of this relation, `curl(1/mu curl A) = J`, rather than the integral form.  This is the equation solved in the `Tesla` miniapp.  Have a closer look at `Tesla` and I'd be happy to offer further advise on how to couple these two mini apps together to obtain your solution.

Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell ,
thanks for the comments!
The question if this approximation of Maxwell's equations holds is somewhat open (the fact that it's frequently used doesn't mean it's valid), but that's also one of the reasons why I wanted to use MFEM rather than commercial software packages, as we can eventually compare different approaches.
As for the calculation of the magnetic potential `A` via `curl(1/mu curl A) = J` (or via the Biot-Savart integral), that's a good point, I will start with your suggestions of coupling the output of `Tesla.cpp` to the a `Volta.cpp`-like approach.

Thanks for the support!
Mathias

**Comment by Maatinski** [comment]
Hey everyone,
a quick follow-up: I have been working on implementing above problem, and (after exploring my brain for residual C++/FEM knowledge) managed to implement it. For now, I simulated the magnetic vector potential in Matlab and imported it via a grid function file `.gf`. For testing, I used a simple spherical object (sigma = 0.3 S/m) with a second off-centered sphere inside (sigma = 0.01 S/m). This model is "excited" by a simple loop at 1000 Hz / 1 Amp coil current. The same setup was simulated in Sim4Life. Even though I have not look at the values numerically (will do that soon), the agreements look good so far. 

**Sim4Life results:**
![test_sim4life](https://user-images.githubusercontent.com/63864061/81014656-40caec80-8e2b-11ea-9a69-f25c514f13a7.png)

**MFEM results:**
![GLVis_s01](https://user-images.githubusercontent.com/63864061/81015458-c00cf000-8e2c-11ea-8efa-1f3fda9d139d.png)

**Parallel version:**
In order to get an idea about the runtime for larger models (up to 200M hex elements), I also turned this into a parallel version (pretty much by replicated what was done in `volta.cpp`). In this case, however, I ran into a problem, which I have not yet been able to resolve. There is some weird artifacts in the electric field maps (primarily located in one half of the domain). This is also the case if I run only a single MPI process:
![GLVis_s01](https://user-images.githubusercontent.com/63864061/81016858-3dd1fb00-8e2f-11ea-8f2b-50bcc6c08002.png)

Do you have an idea where this effect is coming from. I can also clean up my code and post it here, if this is not an obvious mistake I made.

In any case, thanks a lot for your help!
Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
It looks like you've made good progress.  In the parallel case it looks like your field needs to be reordered.  If you're reading the same `.gf` file into both the serial and parallel versions of the code I'm not surprised the results look odd.  In serial MFEM follows the vertex ordering of your original mesh.  The ordering of edges and nodes then follow from the element order but their orientations are determined from the vertex numbering.  In parallel it is assumed that the vertices will need to be renumbered so we generate a new vertex numbering by parsing the local elements.  The new vertex numbering can then effect edges and faces through their orientations.  It is, of course, possible that MFEM will generate the same vertex numbering in both cases but this is highly unlikely.  Sometimes the two orderings will be similar in certain regions which may explain the smooth appearance in the upper half of the domain.  So, generally serial and parallel GridFunctions will not agree.
Best wishes,
Mark

**Comment by Maatinski** [comment]
Thanks, @mlstowell for the quick reply, that explains the behaviour!
I was thinking about implementing the Biot-Savart solver in C++ anyway (i.e., importing the 3D coil path and directly simulating the magn. potentials). I suppose in this case, extracting the vertices from `pmesh` should have the correct order. Is it possible to do this within an MPI process only for the respective vertices of the `pmesh` section and then consolidating the data (the magn. potentials at the hex vertices) between the different MPI tasks?
Of course the alternative is directly solving the magn. potentials using a workflow similar to `tesla.cpp` (which I also want to look at), but our geometries are given by a small number of straight segments, so solving Biot-Savart is pretty straight forward in this case.

**Comment by v-dobrev** [comment]
You should be able to convert a serial (global) `GridFunction` to a `ParGridFunction` using the following constructor: https://github.com/mfem/mfem/blob/0f067d41393943fbabc76fd62a9ed76c8e9a48e5/fem/pgridfunc.hpp#L73-L77

You will need to create the partitioning array outside the `ParMesh` constructor which can be done with a call like this: https://github.com/mfem/mfem/blob/0f067d41393943fbabc76fd62a9ed76c8e9a48e5/mesh/pmesh.cpp#L148

**Comment by Maatinski** [comment]
Hi all,
first, a (delayed) thanks to @mlstowell and @v-dobrev , this worked very well. I took a few steps further and now import a 3D coil curve (consisting of straight segments) and directly calculate the magnetic potential, which works pretty well. I also managed to increase the complexity and run a somewhat complicated head model (13 different tissues/materials), which works well both in serial and parallel (although convergence is still somewhat slow, but I wanted to deal with that later on).

**Result:**
![GLVis_s02](https://user-images.githubusercontent.com/63864061/81747368-fb836c00-9475-11ea-8f63-129b76b7dd10.png)

One issue that seemed trivial in the beginning (but is now causing some headache): how can I relate the electric field solution to the physical locations, i.e., how can I relate these to 

1) _the actual locations of their respective basis functions_ (I assume for `ND_FECollection` bases, these correspond to points on the edges?),
2) _mesh vertices_ (I tried to adapt the following [issue](https://github.com/mfem/mfem/issues/23), but couldn't make it work)
3) and _arbitrary locations_ (although I assume in this case I can use `Mesh::FindPoints`)

After solving this last issue, I will run some tests on larger sections of our body model (full body has ~50M hexahedral elements), and of course post some updates here.

As always, thanks a lot for all helpful comments!
Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
The solution values, degrees of freedom, of a Nedelec representation of a field correspond to path integrals of the electric field.  For example, when using first order fields the degrees of freedom are path integrals of the electric field along an individual edge i.e. the voltage difference between the vertices at the ends of the edge.  These degrees of freedom tell you nothing about the electric field perpendicular to an edge.

If you need vector field values on the face, edge, or vertices of your mesh you will need to perform some sort of averaging of the values in the neighboring elements because Nedelec fields are discontinuous across element boundaries.  If you need values somewhere inside an element these are relatively easy to obtain.  `FindPoints` is certainly one solution but there are other possibilities.  

Extracting individual field values can be done with visualization tools like [VisIt](http://visit.llnl.gov). MFEM also has an interface to `FindPoints` called `get-values` in the `miniapps/tools` subdirectory which can extract moderate numbers of field values.  If you need something more tailored to your application I'd be happy to work with you to find a suitable solution.

Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell  ,
thanks for the quick reply (and for clarifying)! I think in my case, the easiest would be to sample the electric field components at each hexahedron's center-point (where the electric field should be well defined, right?). If I understand correctly, `FindPoints` might be somewhat of an "overkill", since the structure/order of the hexahedral mesh cells (and their edges) are known. Is there a quick way of obtaining these values?
In any case, I will look a little closer into the examples (especially the one in `miniapps/tools`).

Thanks a lot,
Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
Yes, the cell centers should provide good results.  The basic procedure is to loop over all elements and evaluate your `GridFunction`.  The main caveat is that the element order, particularly in parallel, may not be what you expect.  Consequently it's a good idea to print out the locations as well as the field values.  The loop would look like this:
```
int spaceDim = mesh.SpaceDimension();
Vector cntr(spaceDim);
Vector val;
for (int e=0; e<mesh.GetNE(); e++)
{
   const IntegrationPoint & ip = Geometries.GetCenter(mesh.GetElementBaseGeometry(e));
   ElementTransformation * T = mesh.GetElementTransformation(e);
   T->SetIntPoint(ip);
   T->Transform(ip, cntr);
   gridfunc.GetVectorValue(e, ip, val);
   /// Print the location `cntr` and the field value `val` in whatever manner you like
}
```

Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell ,
Thanks for the comments. I actually just found a quick (and hopefully robust) way. Very similar to your approach, and when I am populating a Cartesian grid with the sampled field data, it's actually very straight forward to locate the correct index of these grid points via:

```
// get the center point of the current mesh cell
const Element *cell = mesh->GetElement(mc);
Array<int> vert_indices;
cell->GetVertices(vert_indices);

// assuming the vertices are numbered correctly, we only need two vertices to get the center point
const double *vert_1 = mesh->GetVertex(vert_indices[0]);
const double *vert_2 = mesh->GetVertex(vert_indices[6]);

// check that the two points really determine the center point
if ( vert_1[0] == vert_2[0] or vert_1[1] == vert_2[1] or vert_1[2] == vert_2[2] )
{
	MFEM_ABORT("Cannot determine center point!");
}

// get the center point
Vector cell_center(3);
cell_center[0] = 0.5 * ( vert_1[0] + vert_2[0] );
cell_center[1] = 0.5 * ( vert_1[1] + vert_2[1] );
cell_center[2] = 0.5 * ( vert_1[2] + vert_2[2] );

// determine the correct index in the grid of size (nxx,nyy,nzz)
int sub_id_x = ( cell_center[0] - grid_min[0] ) / step_size[0];
int sub_id_y = ( cell_center[1] - grid_min[1] ) / step_size[1];
int sub_id_z = ( cell_center[2] - grid_min[2] ) / step_size[2];

// get linear index similar to Matlab convention
int lin_idx = sub_id_x + sub_id_y * nxx + sub_id_z * nyy * nxx;
```

With some tweaks, this should also work for finer Cartesian grids and arbitrary points (when recomputing the grid index <-> mesh cell index links).
Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
Just be warned that in parallel the element and vertex numberings will be unpredictable.  They will be deterministic but they will be based on a non-Cartesian partitioning produced by METIS (unless you provide your own partitioning, I suppose).
Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell ,
thanks for the remark! My feeling is that since I am iterating over the cells and set all the electric field values sequentially this should work, but I will keep this remark in mind, in case I run into trouble. I actually just finished all bits and pieces, including import/export. 
Attached is the first field map in a reasonable portion of our model (6M mesh cells, "excited" by a 1A / 1000Hz coil loop around the head). Convergence with the Hyper AMG precon. worked pretty well.
Again, thanks for providing this great library to everyone, really really impressive!!

![image](https://user-images.githubusercontent.com/63864061/82084995-4beb0b80-96ba-11ea-8224-b851136c9851.png)

**Comment by tzanio** [comment]
Thanks for the good words @Maatinski.

Please let us know if we can help in any other way in the future... and please keep us updated on your results, they look really cool!

Tzanio

**Comment by Maatinski** [comment]
Hey everyone,
one topic did come up today:
I am currently solving the EM simulation in parallel using the AMG preconditioner. I noticed that some boundary cells show very high electric field value. These mesh cells correspond to cells at the boundary of two or three of the parallel mesh partitions. The hot-spots do not occur when solving the exact same problem with only one MPI process.
See attached the screenshot of the model, that shows two high-value mesh cells in the neck area.
Any idea where this is coming from?
Thanks,
Mathias

![GLVis_s01](https://user-images.githubusercontent.com/63864061/82483105-dca05d80-9aa5-11ea-8969-6c28b49efa46.png)

**Comment by mlstowell** [comment]
That's odd. Do the solutions look the same aside from these two hotspots?  How large are these hotspots relative to the other field values?

**Comment by Maatinski** [comment]
Yes, otherwise they look very similar, but the hot-spots can be pretty significant (10-fold higher than the rest). If this is not a trivial issue, I will run some more test and compare the exported field maps when run with different numbers of MPI processes...
Maatinski

**Comment by mlstowell** [comment]
It could be an indexing error but since the solutions look very similar this seems unlikely.  It could be that the `GridFunction` potential is not being set properly from the solution vector.  Or it could be something with computing `E` from `phi`.  Does `phi` look reasonable?

I'd be happy to look at your code if you think that might help.

**Comment by Maatinski** [comment]
Hi @mlstowell ,
I've had some time today to look into this and did some testing on a simpler model (cube with a smaller cube inside, i.e., two materials). One interesting thing is that these errors correspond to the "index=0" mesh cells of each MPI partition. Other than that I haven't been able to resolve this.
Attached are the relevant lines of code and some screenshots of the respective `GridFunction`.

**Mesh preparation:**
```
int part_method = 1;
int *partitioning = mesh->GeneratePartitioning(num_procs, part_method);
pmesh = new ParMesh(MPI_COMM_WORLD, *mesh, partitioning );
delete mesh;

// Ensure that quad and hex meshes are treated as non-conforming.
pmesh->EnsureNCMesh();
```

**Calculate magnetic vector potential:**
```
int num_query_points = pmesh->GetNV();
Vector magn_pot(3*num_query_points,0);
double query_point[3];
for ( int qid=0; qid<num_query_points; qid++ )
{
	// get current query point
	pmesh->GetNode( qid, query_point);



	magn_pot[qid] = Asum_vec[0];
	magn_pot[qid+num_query_points] = Asum_vec[1];
	magn_pot[qid+2*num_query_points] = Asum_vec[2];

		}
```

**Comment by mlstowell** [comment]
Hi, @Maatinski ,

Given your findings regarding element 0 on each processor, the most likely culprit is `ess_bdr_todfs`.  What boundary condition are you applying on the `DivSigmaGrad_` solve and how are you setting `ess_bdr_tdofs`?

Another thing, which probably is not a problem, is that `DivA_src` is not technically a `ParGridFunction`.  It would be more proper to say that it is a `ParLinearForm`.  The drawback is that you cannot visualize these.  However, if you manually compute the divergence of `A` I suspect it won't match your `DivA_src` plot in magnitude.  `DivA_src` may agree with the true divergence qualitatively (since you're using a Cartesian mesh) but the magnitude is probably scaled by some factor of the element volume.

The only other thing that gives me pause is that `A_src` is in H1^3 rather than HCurl.  This implies that `e_` is also in H1^3 since you add `A_src` to `e_` after computing `GradPhi`.  This might be a fine choice but I'm just a little wary of using the `grad` operator to compute `e_` in this basis because I don't think the `grad` operator (how do you compute this?) performs the necessary nodal averages to produce a reliable `e_`.

Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell ,
thanks so much for the comprehensive comments!
I changed my previous comment and added all the definitions of the gird function, bi/linear forms, operators, etc. Most of them are similar to the `Volta.cpp` example.
As for the boundary condition: you're absolutely right! I am not setting specific Dirichlet or Neumann boundary conditions, but just remove the first DoF on processor one (again identical to `Volta.cpp`) which also led to very good agreement with the commercial solution we have (Sim4Life). I did, however, forgot the `myid==0` condition, so this step happened on every processor. This solved the issue with the hot-spots.
As for your other comments: I will definitely look into this. If the code lines I added above reveal a mistake or point that can be improved, I am of course happy for any hints!

Thanks and warm regards,
Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
It all looks good to me.  I'm glad to hear you sorted out the boundary condition issue.  I expect your visualization of `DivA_src` is just for debugging purposes but if you want to visualize the properly scaled field you can obtain this by performing a mass matrix solve (which is relatively inexpensive):
```
ParGridFunction DivA_gf(H1FEspace);
DivA_gf = 0.0;
{
   ParBilinearForm h1Mass(H1FEspace);
   h1Mass.AddDomainIntegrator(new MassIntegrator);
   
   HypreParMatrix MassH1;
   Vector DivA_sol, DivA_rhs;

   Array<int> dbc_dofs_h1(0);
   h1Mass.FormLinearSystem(dbc_dofs_h1, DivA_gf, *DivA_src, MassH1, DivA_sol, DivA_rhs);
   HyprePCG pcgM(MassH1);
   pcgM.SetTol(1e-12);
   pcgM.SetMaxIter(500);
   pcgM.SetPrintLevel(0);
   HypreDiagScale diagM;
   pcgM.SetPreconditioner(diagM);
   pcgM.Mult(DivA_rhs, DivA_sol);

   h1Mass.RecoverFEMSolution(DivA_sol, *DivA_src, DivA_gf);
}
```
... or something like that.

By the way, @tzanio showed some of your images from this issue to the MFEM team in last week's team meeting.  People are excited to learn more about your project and see more great images!

Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell ,
as always thanks a lot for the comments! As you correctly expected, I only used the visualization of `DivA_src` for debugging, but I will keep your hint in mind. I have been successfully running the first simulations in our full model (50M mesh cells), seems to work very well. Since we intend to run a large number of simulations in the same mesh model (using different coils), I modified the code to avoid redundant steps (everything related to mesh preparation and assembly of the bi/linear forms is done once in the very beginning). 
I was wondering, if there's also a way to also "reuse" the AMG preconditioner (or anything related to this step) for different `DivA_src` source terms? My feeling is that the AMG "assembly" takes longer than the actual solve of the linear system.

Finally, thanks a lot for the interest in our modeling work. I will reach out to you via mail to send some more details on what exactly we're using MFEM for.

Maatinski

**Comment by mlstowell** [comment]
Hi, @Maatinski ,
In theory I believe it is possible to reuse much of the work done coarsening the matrix in AMG with multiple right hand sides.  To be honest I'm not sure if HYPRE provides a convenient means to do this or if that interface is available through MFEM.  @tzanio or @barker29 would probably know much more about that than I do.

Glad to hear things are going well.

Best wishes,
Mark

**Comment by tzanio** [comment]
Hi @Maatinski,

> I was wondering, if there's also a way to also "reuse" the AMG preconditioner (or anything related to this step) for different DivA_src source terms? 

The AMG preconditioner depends only on the matrix, so you should absolutely reuse it of only the right-hand side changes.

You can also try reusing the same AMG preconditioner when the matrix changes, but that may not work well if the change is large.

Hop this helps,
Tzanio

**Comment by tzanio** [comment]
Hi @Maatinski, I will close this issue for now, but please feel free to reopen if you have further questions. -Tzanio

**Comment by mlstowell** [comment]
Hi, @Maatinski, 

I don't have much experience producing and loading non-conforming meshes but I believe it can be done. The meshes do require a special format which is described here: [NC Mesh v1.0](https://mfem.org/mesh-formats/#mfem-nc-mesh-v10). The real expert with such things is @jakubcerveny. If you're following this format and still running into trouble I'm sure we would want to fix or clarify things at our end. Please let us know what sorts of problems you might encounter.

Best wishes,
Mark

**Comment by Maatinski** [comment]
Perfect, that's very helpful! I will give it a try with this file format; seems pretty straight forward. I assume that when the mesh is imported using this format, MFEM knows how to handle the rest (such as identifying tdofs, etc.).
At the moment I am using my own file format (in binary format) since otherwise the ASCII mesh files get very very big, but I assume manually generating an NCMesh should also be possible.

Mathias

**Comment by Maatinski** [comment]
Hi all,
a quick follow-up on the definition of _hanging nodes_. I did look through some documentation (of course the MFEM documentation), but all examples I could find are always in 2D.
Let's day I have a uniform (conforming) 3D hexahedral mesh. A single edge in this mesh (somewhere in the middle) is shared by a total of four mesh cells. If I split only one of these four cells this creates a hanging node lying on that edge.
My question is: would this node also be classified as a hanging node, if the neighboring mesh cell (sharing a facet with the first cell) is also split (i.e. a total of 2 cells are split)? Or asked more generally: is this node classified as hanging node if and only if 1, 2, or 3 of these 4 mesh cells are split.
I realize this question might be difficult to grasp without any visualization; I am happy to make a simple drawing, if that helps. 

As always thanks a lot for the help and warm regards,
Maatinski

**Comment by Maatinski** [comment]
Hi all,
another question popped up today. I am pretty much done generating the non-conforming meshes for our body models, but I am still a little unclear about the NCmesh format, especially the `vertex_parent` section. Taking [amr-hex.mesh](https://github.com/mfem/mfem/blob/master/data/amr-hex.mesh) as an example:
Here, there is only 8 top-level vertices; all other vertices can be computed from the `vertex_parent` list. Does this mean that all vertices listed in `vertex_parent` are hanging nodes? Wouldn't this leave the FEM problem with virtually no degrees-of-freedom to obtain the solution? Otherwise I may be misinterpreting the explanation: 

[mfem-nc-mesh](https://mfem.org/mesh-formats/#mfem-nc-mesh-v10)
_"The nonconfoming mesh however needs to identify hanging vertices, which may occur in the middle of edges or faces as elements are refined. A hanging vertex always has two parent vertices. ... In our example above, vertices 6, 7, 8, 9 are hanging vertices, so the vertex_parents section would look like this:"_

In other words, couldn't there be non-top-level vertices that are not classified as hanging nodes?

Thanks a lot
Maatinski

**Comment by tzanio** [comment]
@jakubcerveny can you help?

**Comment by jakubcerveny** [comment]
Hi @Maatinski,

you are right that the description of the `vertex_parents` section may be oversimplified. The section contains all vertices that have been derived as mid-points of some segments. Not all of them are necessarily truly hanging (as in a uniformly refined mesh), but all of them potentially are, e.g. if some elements get derefined later. You may think of the `vertex_parents` section as part the refinement history, only on vertices. 

Does that make sense? Do you think it would help if we reformulated the description on the web?

Jakub

**Comment by Maatinski** [comment]
Hi Jakub,
thanks for the clarification. Yes, I think in that case the explanation could maybe be reformulated. I think I was also slightly confused by the example sketch, in which all children vertices happen to be hanging nodes. Maybe changing this to a mesh where this is not the case would be helpful.

![image](https://user-images.githubusercontent.com/63864061/118146792-f2a4dd80-b3dc-11eb-8147-9bb48f2a5d06.png)

Thanks a lot and warm regards,
Maatinski

**Comment by Maatinski** [comment]
Looks like it's working. The solutions look fairly similar, despite the ~6-fold reduction in number of mesh cells. The mesh is not optimized yet (i..e, no error-based refinement); it's simply generated based on the voxelmodel with some grading constraints and then imported by MFEM (no further refinement within MFEM).

Meshes
![image](https://user-images.githubusercontent.com/63864061/118335218-27985980-b4dd-11eb-9197-79f51569d808.png)

Solutions
![image](https://user-images.githubusercontent.com/63864061/118334967-9aed9b80-b4dc-11eb-8ebe-a5679ef4ec25.png)

Zoomed
![image](https://user-images.githubusercontent.com/63864061/118335257-4139a100-b4dd-11eb-9f31-3eb71d76e2f8.png)

There is also some "stripe" artifact in the NC mesh solution (left), will look into this.
But for now, this functionality seems great! We'll definitively start using this in our modeling.

Maatinski

**Comment by Maatinski** [comment]
A quick question about these almost "stripe-artefacts":
![image](https://user-images.githubusercontent.com/63864061/118543191-f8bef500-b721-11eb-863c-6ace4454b3a0.png)

I assume it's difficult to tell for sure from these screenshots, but does this look like it's only due to the coarseness of the mesh in this area, or is there a chance that the vertex parent definitions are incorrect in the mesh? There is no change in conductivity in this area, so I am a little surprised by the lack of smoothness of the solution

Thanks,
Maatinski

**Comment by tzanio** [comment]
Hi @Maatinski,

What type of field are we looking at here?

For an H1 function, the solution should be continuous (with the values coming from the coarser side).

If it is the magnitude of an H(curl) field then that could be generally discontinuous, though the stripes do look a bit suspect.

Tzanio

**Comment by jakubcerveny** [comment]
@Maatinski, could there be a CPU boundary between the coarse and fine faces?

**Comment by Maatinski** [comment]
Hi all,
I have finally been able to look into this:
@tzanio It's an H(curl) field, so I assume the discontinuity might not necessarily be an artifact.
@jakubcerveny This simulation was done in serial, i.e., no CPU boundaries.

I redid some testing with further grading the mesh (essentially having one more refinement iteration, where every cell neighboring a cell with larger size is also split). In this case, these stripe artifacts are completely gone, and the agreement with the full solution is pretty good. Maybe the transition from fine to coarse cells was simply too abrupt?

Maatinski

![image](https://user-images.githubusercontent.com/63864061/118989188-d4486000-b94f-11eb-8ece-bc5aa9fcb800.png)

![image](https://user-images.githubusercontent.com/63864061/118989207-d8747d80-b94f-11eb-9b2c-46b46caa1266.png)
